File: prove_llm_code/max.cpp
#include <iostream>
#include <z3++.h>

int main() {
    z3::context c;
    z3::solver s(c);

    z3::expr a = c.int_const("a");
    z3::expr b = c.int_const("b");
    z3::expr r = c.int_const("r");

    s.add(r == z3::ite(a >= b, a, b));

    z3::expr bug = (r < a) || (r < b) || ((r != a) && (r != b));
    s.add(bug);

    z3::check_result result = s.check();

    std::cout << "Solver result: " << result << "\n";

    if (result == z3::sat) {
        std::cout << "Counterexample found:\n" << s.get_model() << "\n";
    } else if (result == z3::unsat) {
        std::cout << "No counterexample exists. max satisfies the spec.\n";
    } else {
        std::cout << "Solver returned unknown.\n";
    }

    return 0;
}

File: prove_llm_code/max.py
from z3 import Int, If, Or, And, Solver, sat, unsat

# Symbolic inputs
a = Int("a")
b = Int("b")

# Symbolic return value
r = Int("r")

s = Solver()

# Encode:
# int max2(int a, int b) {
#     if (a >= b) return a;
#     else return b;
# }
s.add(r == If(a >= b, a, b))

# Ask for a counterexample:
# max2 is wrong if:
# 1) r < a, or
# 2) r < b, or
# 3) r is neither a nor b
bug = Or(
    r < a,
    r < b,
    And(r != a, r != b)
)

s.add(bug)

result = s.check()
print("Solver result:", result)

if result == sat:
    print("Counterexample found:")
    print(s.model())
elif result == unsat:
    print("No counterexample exists. max2 satisfies the spec.")
else:
    print("Solver returned unknown.")